workshop1/slides/src/slides.md

```diff
@@ -0,0 +1,664 @@
+parseLine :: CSVLine -> Maybe Candidate
+parseLine line = makeCandidate (splitOn ',' line)
+
+makeCandidate :: [String] -> Maybe Candidate
+makeCandidate [c, p, s, a] =
+  fromMaybeAge (Candidate c p s) (readMaybe a)
+makeCandidate _ = Nothing
+
+fromMaybeAge
+  :: (Double -> Candidate)
+  -> Maybe Double
+  -> Maybe Candidate
+fromMaybeAge partialCandidate Nothing  = Nothing
+fromMaybeAge partialCandidate (Just a) = Just (partialCandidate a)
+```
+
+
+## Stripping Quotes
+
+```haskell
+stripQ :: String -> String
+stripQ = undefined
+  where
+    leftStrip :: String -> String
+    leftStrip ('"' : xs) = xs
+    leftStrip xs         = xs
+
+    rightStrip :: String -> String
+    rightStrip = undefined
+```
+
+* Use `leftStrip` and `reverse` to implement `rightStrip`
+* Use `leftStrip` and `rightStrip` to implement `stripQ`
+* `reverse` reverses items in a list
+* `where` used to define local functions
+
+## Stripping Quotes
+
+Our solution:
+```haskell
+stripQ :: String -> String
+stripQ = leftStrip . rightStrip
+  where
+    leftStrip :: String -> String
+    leftStrip ('"' : xs) = xs
+    leftStrip xs         = xs
+
+    rightStrip :: String -> String
+    rightStrip = reverse . leftStrip . reverse
+```
+
+* Point-free style: composition FTW!
+
+## Improving `parseLine`
+
+* Using `stripQ` to improve `makeCandidate`
+
+```haskell
+...
+
+makeCandidate :: [String] -> Maybe Candidate
+makeCandidate [c, p, s, a] =
+  let c' = stripQ c
+      p' = stripQ p
+      s' = stripQ s
+  in fromMaybeAge (Candidate c' p' s') (readMaybe a)
+makeCandidate _ = Nothing
+
+...
+```
+
+* `let ... in ...` binding useful to split up functions
+
+
+## The `map` function
+
+* How do we apply `parseLine` to every row?
+* Use `map`, a *higher-order function*
+  * Takes two inputs: a function and a list.
+  * Applies the function to every item in the list
+
+```haskell
+map :: (a -> b) -> [a] -> [b]
+map f []     = []
+map f (x:xs) = f x : map f xs
+```
+
+* Defined using *recursion*
+
+
+## Writing `parseFile :: CSV -> DataModel`
+
+* We are ready to write the full parser function:
+
+```haskell
+type CSV = String
+type DataModel = [Maybe Candidate]
+
+parseFile :: CSV -> DataModel
+parseFile = undefined
+```
+
+* `lines` splits strings into lines
+* `tail` removes the first item in a list
+* Try and use point-free style
+
+## Writing `parseFile :: CSV -> DataModel`
+
+Our solution:
+
+```haskell
+parseFile :: CSV -> DataModel
+parseFile = map parseLine . tail . lines
+```
+
+
+## Generating `encode :: DataModel -> JSON`
+
+```haskell
+{-# LANGUAGE DeriveGeneric #-}
+main Main where
+
+import           GHC.Generics
+import qualified Data.Aeson      as JSON
+import qualified Data.ByteString as B
+
+data Candidate = Candidate
+  { ...
+  } deriving (Show, Generic)
+
+instance Aeson.ToJSON Candidate
+```
+
+* Language extensions expand the Haskell language
+* `Data.Aeson` provides encoding to and from JSON
+* `Data.ByteString` provides efficient binary strings
```

Comment:
`Data.ByteString.Lazy` by default

Reply:
Why? If Lazy should be the default, don't you think the ByteString authors would have *made* it the default? It doesn't do what most people assume it does, anyway.

Reply:
Ah thanks

Reply:
> Why? If Lazy should be the default, don't you think the ByteString authors would have _made_ it the default?

I think he was referring to how `Data.ByteString.Lazy` is the default for the `encode` function in `Data.Aeson`

Reply:
@chreekat

[encode creates a lazy bytestring](http://hackage.haskell.org/package/aeson-1.4.2.0/docs/Data-Aeson.html#v:encode)